Document features (#1257)
software-mansion/scarb: website/docs/reference/features.md

Author: Yeah, maybe it's a good idea, because it isn't that big of a feature :) and wouldn't be used that much to have it's own "main" page. I'll move it there.

Author: By a sub section, you mean a separate page but grouped with Conditional Compilation (like Starknet pages in Core extensions) or just putting it on one page underneath the text of Conditional Compilation?

Reviewer: Hi!
So sorry for late response!
The first one, or even on the same page.

# Features

Features in Scarb provide a way to conditionally compile specific parts of the code during the build process.

## `[features]` section

> In the `Scarb.toml` file, features are defined within the `[features]` section. Each feature lists other features that it enables.

Reviewer: ```suggestion
A package defines a set of named features in the `[features]` section of `Scarb.toml` file. Each defined feature can list other features that should be enabled with it.
```

For example, a package supporting various hash functions might define features like this:

```toml
[features]
poseidon = []
pedersen = []
keccak = []
```

> With these features set, conditional compilation expressions (`cfg` expressions) can selectively include code to support requested features during compile time. For instance:

Reviewer: ```suggestion
With these features set, conditional compilation (`cfg`) attributes can be used to selectively include code to support requested features during compile time. For instance:
```

```rust

> // Conditionally include a package

Reviewer: ```suggestion
// Conditionally include a module
```

#[cfg(feature: 'poseidon')]
mod poseidon;

// Conditionally define a function
#[cfg(feature: 'pedersen')]
fn hash_pedersen(value: felt252) -> felt252 {
  // ...
}
```

To enable specific features, use the `--features` flag followed by a comma-separated list of features. For example, to build with only the `poseidon` and `pedersen` features enabled:

```
scarb build --features poseidon,pedersen
```

Enabling all features can be done with the `--all-features` flag.

## `default` features

> By default, all features are disabled unless explicitly enabled with the `--features` flag. However, this behavior can be changed by specifying a default feature in the `[features]` section, like so:

Reviewer: ```suggestion
By default, all features are disabled unless explicitly enabled. However, this behaviour can be changed by specifying a default feature in the `[features]` section, like so:
```

```toml
[features]
default = ["poseidon", "pedersen"]
poseidon = []
pedersen = []
keccak = []
```

> During compilation, the compiler will enable the default feature, which in turn activates all listed features.

Reviewer: ```suggestion
When the package is built, the default feature is enabled which in turn enables the listed features.
```
